src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,468 @@
+/// witness Vec<StackElement>. Does not interpret/check the witness against the miniscript inside
+/// the descriptor
+fn verify_wsh<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &bitcoin::Script,
+    witness: &'txin [Vec<u8>],
+) -> Result<(Miniscript<PublicKey, hash160::Hash>, Vec<StackElement<'txin>>), Error>
+{
+    if *script_sig != Script::new(){
+        return Err(Error::BadDescriptor)
+    }
+    if let Some((witness_script, witness)) = witness.split_last() {
+        let witness_script = Script::from(witness_script.clone());
+        if witness_script.to_v0_p2wsh() != *script_pubkey {
+            return Err(Error::BadDescriptor)
+        }
+        let ms = Miniscript::parse(&witness_script)?;
+        //only iter till len -1 to not include the witness script
+        let stack: Vec<StackElement> = witness.iter()
+            .map(|elem| StackElement::from(elem)).collect();
+        Ok((ms, stack))
+    } else{
+        Err(Error::BadDescriptor)
+    }
+}
+
+
+/// Creates a pkh descriptor based on scriptsig and script_pubkey. Validates the hash checks for
+/// p2pkh against top element(pk) and pushes all remaining witness elements into witness<StackElement>
+fn verify_p2pkh<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &'txin bitcoin::Script,
+    witness: & [Vec<u8>],
+) -> Result<(Descriptor<PublicKey, hash160::Hash>, Vec<StackElement<'txin>>), Error>
+{
+    let (pk_bytes, stack) = parse_scriptsig_top(script_sig)?;
+    if let Ok(pk) = bitcoin::PublicKey::from_slice(&pk_bytes) {
+        let addr = bitcoin::Address::p2pkh(
+            &pk.to_public_key(),
+            bitcoin::Network::Bitcoin,
+        );
+        if *script_pubkey != addr.script_pubkey() || !witness.is_empty() {
+            return Err(Error::BadDescriptor)
+        }
+        Ok((Descriptor::Pkh(pk), stack))
+    } else{
+        Err(Error::BadDescriptor)
+    }
+}
+
+/// Helper for creating a p2sh descriptor based on script_pubkey and witness. Validates the p2sh
+/// hash based on redeem script, pops the redeem script from the script sig stack, translates other
+/// elements from scriptsig into Vec<StackElement>
+fn verify_p2sh<'txin>(
+    script_pubkey: &bitcoin::Script,
+    script_sig: &'txin bitcoin::Script,
+) -> Result<(Script, Vec<StackElement<'txin>>), Error>
+{
+    let (redeem_script, stack) = parse_scriptsig_top(script_sig)?;
+    let redeem_script = Script::from(redeem_script);
+    if redeem_script.to_p2sh() != *script_pubkey {
+        return Err(Error::BadDescriptor)
+    }
+    Ok((redeem_script, stack))
+}
+
+/// Figures out the the type of descriptor based on scriptpubkey, witness and scriptsig.
+/// Outputs a descriptor and witness_stack for all descriptors which can be directly fed into the
+/// interpreter. All script_sig and witness are translated into a single witness stack
+/// Vec<StackElement>
+/// 1) PK: creates a Pk descriptor and transfers the scriptsig to witness_stack<StackElement>
```

Comment:
Can you put `Pk`, `Pkh`, etc in backticks? It makes the documentation a bit easier to read.

I'd also prefer you use `1.` `2.` etc rather than `1)` `2)` etc, but I don't feel strongly about it. cargo doc does the right thing regardless.